### PR TITLE
feat: Implement site banner API and UI - EXO-65534 - Meeds-io/MIPs#72

### PR DESCRIPTION
**Pull request by GouadriaHanen**

### CODE REVIEW

File: component/service/src/main/java/org/exoplatform/social/rest/api/RestUtils.java [hunk 1]
```diff
@@ -345,4 +358,60 @@ public static IdentityManager getIdentityManager() {
     }
     return identityManager;
   }
+
+  public static String getBaseURLSiteRest() {
+    return "/" + PortalContainer.getCurrentPortalContainerName() + "/" + PortalContainer.getCurrentRestContextName()
+        + BASE_URL_SITE_REST_API;
+  }
+
+  public static String buildAttachmentUrl(String siteName, String type) {
```
Comment (azayati) [inline component/service/src/main/java/org/exoplatform/social/rest/api/RestUtils.java]: why not adding that here https://github.com/Meeds-io/social/blob/84be6394c4a559220ef01c6b26e421b73ff0acac/component/core/src/main/java/org/exoplatform/social/core/service/LinkProvider.java#L47 ?

File: component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java [hunk 2]
```diff
@@ -135,13 +155,81 @@ public Response getSites(@Context
...
+    } catch (ObjectNotFoundException | IOException e) {
+      return Response.status(Response.Status.NOT_FOUND).build();
+    }
+  }
+
+  private InputStream getDefaultBannerInputStream(String siteName) throws IOException {
+    byte[] defaultSiteBanner = null;
+    InputStream is =  portalContainer.getPortalContext()
+                   .getResourceAsStream(System.getProperty("sites." + siteName + ".defaultBannerPath",
+                                                           "/images/sites/" + siteName.toLowerCase() + ".png"));
```
Comment (azayati) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java]:
```suggestion
                                                           "/images/sites/banner/" + siteName.toLowerCase() + "Banner.png"));
```

File: component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java [hunk 3]
```diff
@@ -135,13 +155,81 @@ public Response getSites(@Context
...
+
+      return builder.build();
+    } catch (ObjectNotFoundException | IOException e) {
+      return Response.status(Response.Status.NOT_FOUND).build();
+    }
+  }
+
+  private InputStream getDefaultBannerInputStream(String siteName) throws IOException {
+    byte[] defaultSiteBanner = null;
+    InputStream is =  portalContainer.getPortalContext()
```
Comment (azayati) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java]:
```suggestion
    InputStream is = portalContainer.getPortalContext()
```

File: component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java [hunk 4]
```diff
@@ -135,13 +155,81 @@ public Response getSites(@Context
...
+      builder.cacheControl(CACHECONTROL);
+      builder.expires(new Date(System.currentTimeMillis() + CACHE_IN_MILLI_SECONDS));
+
+      return builder.build();
+    } catch (ObjectNotFoundException | IOException e) {
+      return Response.status(Response.Status.NOT_FOUND).build();
+    }
+  }
+
+  private InputStream getDefaultBannerInputStream(String siteName) throws IOException {
```
Comment (azayati) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java]:
```suggestion
  private InputStream getDefaultSiteBannerInputStream(String siteName) throws IOException {
```

File: component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java [hunk 5]
```diff
@@ -135,13 +155,81 @@ public Response getSites(@Context
...
+    InputStream is =  portalContainer.getPortalContext()
+                   .getResourceAsStream(System.getProperty("sites." + siteName + ".defaultBannerPath",
+                                                           "/images/sites/" + siteName.toLowerCase() + ".png"));
+    if (is == null) {
+      is = portalContainer.getPortalContext().getResourceAsStream(SITE_DEFAULT_BANNER_URL);
+      defaultSiteBanner = is != null ? IOUtil.getStreamContentAsBytes(is) : new byte[] {};
+    } else {
+      defaultSiteBanner = IOUtil.getStreamContentAsBytes(is);
+    }
+    return new ByteArrayInputStream(defaultSiteBanner);
```
Comment (azayati) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java]: why not returning directly is ?

File: component/service/src/test/java/org/exoplatform/social/rest/impl/site/SiteRestTest.java [hunk 6]
```diff
@@ -20,6 +20,16 @@
...
 
+import org.exoplatform.container.PortalContainer;
+import org.exoplatform.portal.jdbc.entity.ContainerEntity;
+import org.exoplatform.portal.jdbc.entity.WindowEntity;
+import org.exoplatform.portal.mop.dao.ContainerDAO;
+import org.exoplatform.portal.mop.dao.WindowDAO;
+import org.exoplatform.portal.mop.rest.NavigationRest;
+import org.exoplatform.portal.mop.service.LayoutService;
+import org.exoplatform.services.security.IdentityConstants;
+import org.exoplatform.social.rest.api.RestUtils;
```
Comment (azayati) [inline component/service/src/test/java/org/exoplatform/social/rest/impl/site/SiteRestTest.java]: remove useless imports

File: component/service/src/test/java/org/exoplatform/social/rest/impl/site/SiteRestTest.java [hunk 7]
```diff
@@ -95,18 +124,76 @@ public void testGetSiteNavigation() throws Exception {
...
+    siteEntity.setSiteBody("[{\"children\":[],\"id\":" + bodyId + ",\"type\":\"CONTAINER\"}]");
     siteEntity.setSiteType(siteType);
     siteEntity.setSkin("testSkin");
     siteEntity.setProperties("{}");
     siteEntity.setDisplayed(displayed);
     siteEntity.setDisplayOrder(displayOrder);
     return siteDAO.create(siteEntity);
   }
+
+  private ContainerEntity createContainerInstance() {
```
Comment (azayati) [inline component/service/src/test/java/org/exoplatform/social/rest/impl/site/SiteRestTest.java:201]: why do we need ContainerEntity ?

**Review by azayati — CHANGES_REQUESTED** [review]

File: component/core/src/main/java/org/exoplatform/social/core/service/LinkProvider.java [hunk 8]
```diff
@@ -588,4 +592,61 @@ public static String getResourceBundleLabel(Locale locale, String label) {
     ResourceBundleService resourceBundleService =  ExoContainerContext.getService(ResourceBundleService.class);
     return resourceBundleService.getResourceBundle(resourceBundleService.getSharedResourceBundleNames(), locale).getString(label);
   }
+
+  public static String getBaseURLSiteRest() {
+    return "/" + PortalContainer.getCurrentPortalContainerName() + "/" + PortalContainer.getCurrentRestContextName()
+        + BASE_URL_SITE_REST_API;
+  }
+
+  public static String buildAttachmentUrl(String siteName, String type, boolean isDefault) {
```
Comment (azayati) [inline component/core/src/main/java/org/exoplatform/social/core/service/LinkProvider.java]:
```suggestion
  public static String buildSiteBannertUrl(PortalConfig site) {
```
try to call the private existing method buildAttachmentUrl with the needed site params

File: component/core/src/main/java/org/exoplatform/social/core/service/LinkProvider.java [hunk 9]
```diff
@@ -588,4 +592,61 @@ public static String getResourceBundleLabel(Locale locale, String label) {
...
+                                                  .append(type)
+                                                  .append("?r=")
+                                                  .append(token)
+                                                  .append("&isDefault=")
+                                                  .append(isDefault)
+                                                  .toString();
+
+  }
+
+  public static String generateAttachmentToken(String siteName, String attachmentType) {
```
Comment (azayati) [inline component/core/src/main/java/org/exoplatform/social/core/service/LinkProvider.java]: IMO we can use the already existing method generateAttachmentToken

File: component/core/src/main/java/org/exoplatform/social/core/service/LinkProvider.java [hunk 10]
```diff
@@ -588,4 +592,61 @@ public static String getResourceBundleLabel(Locale locale, String label) {
...
+        token = codecInitializer.getCodec().encode(tokenPlain);
+      } catch (TokenServiceInitializationException e) {
+        LOG.warn("Error generating token of {} for site {}. An empty token will be used", attachmentType, siteName, e);
+        token = StringUtils.EMPTY;
+      }
+    }
+    return token;
+  }
+
+  public static boolean isAttachmentTokenValid(String token, String siteName, String attachmentType) {
```
Comment (azayati) [inline component/core/src/main/java/org/exoplatform/social/core/service/LinkProvider.java]: IMO we can use the already existing method isAttachmentTokenValid

File: component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java (hunk 4 above)
Comment (azayati) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java]: can you consider this https://github.com/Meeds-io/social/pull/2851/files#r1326055319 ?

File: component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java (hunk 5 above)
Comment (azayati) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java]: can you consider this https://github.com/Meeds-io/social/pull/2851/files#r1326065689 ?

File: component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java [hunk 11]
```diff
@@ -187,7 +203,57 @@ public Response getSiteById(@Context
...
   }
 
+  @GET
+  @Path("{siteName}/banner")
+  @RolesAllowed("users")
+  @Produces("image/png")
+  @Operation(summary = "Gets a site banner", method = "GET")
+  @ApiResponses(value = { @ApiResponse(responseCode = "200", description = "Request fulfilled"),
+      @ApiResponse(responseCode = "500", description = "Internal server error"),
+      @ApiResponse(responseCode = "400", description = "Invalid query input"),
```
Comment (azayati) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java]: I think we don't have the 400 error

File: component/core/src/main/java/org/exoplatform/social/core/service/LinkProvider.java [hunk 12]
```diff
@@ -89,6 +89,10 @@ public class LinkProvider {
 
   public static final String TYPE = "file";
 
+  public static final String     BASE_URL_SITE_REST_API   = "/v1/social/sites";
+
+  public static final String     ATTACHMENT_BANNER_TYPE   = "banner";
```
Comment (azayati) [inline component/core/src/main/java/org/exoplatform/social/core/service/LinkProvider.java]:
```suggestion
  public static final String     SITE_BANNER_TYPE   = "banner";
```

File: component/service/src/test/java/org/exoplatform/social/rest/impl/site/SiteRestTest.java (hunk 7 above)
Comment (azayati) [reply component/service/src/test/java/org/exoplatform/social/rest/impl/site/SiteRestTest.java:201]: can you consider this https://github.com/Meeds-io/social/pull/2851/files#r1326083267 ?

**Review by azayati — CHANGES_REQUESTED** [review]

**Comment by exo-swf** [comment]
Your PR triggers too many exo-ci builds! Please finish your work and then, set your PR ready! Thank you

**Review by azayati — APPROVED** [review]
LGTM

**Pull request merged**